crates/ruff_benchmark/src/real_world_projects.rs

```diff
@@ -0,0 +1,425 @@
+    pub commit: &'a str,
+    /// List of paths within the project to check (`ty check <paths>`)
+    pub paths: Vec<&'a SystemPath>,
+    /// Dependencies to install via uv
+    pub dependencies: Vec<&'a str>,
+    /// Limit candidate packages to those that were uploaded prior to a given point in time (ISO 8601 format).
+    /// Maps to uv's `exclude-newer`.
+    pub max_dep_date: &'a str,
+    /// Python version to use
+    pub python_version: PythonVersion,
+}
+
+impl<'a> RealWorldProject<'a> {
+    /// Setup a real-world project for benchmarking
+    pub fn setup(self) -> Result<SetupProject<'a>> {
+        let start = Instant::now();
+        tracing::debug!("Setting up project {}", self.name);
+
+        // Create project directory in cargo target
+        let project_root = get_project_cache_dir(self.name)?;
+
+        // Clone the repository if it doesn't exist, or update if it does
+        if project_root.exists() {
+            tracing::debug!("Updating repository for project '{}'...", self.name);
+            let start = std::time::Instant::now();
+            update_repository(&project_root, self.commit)?;
+            tracing::debug!(
+                "Repository update completed in {:.2}s",
+                start.elapsed().as_secs_f64()
+            );
+        } else {
+            tracing::debug!("Cloning repository for project '{}'...", self.name);
+            let start = std::time::Instant::now();
+            clone_repository(self.repository, &project_root, self.commit)?;
+            tracing::debug!(
+                "Repository clone completed in {:.2}s",
+                start.elapsed().as_secs_f64()
+            );
+        }
+
+        let checkout = Checkout {
+            path: project_root,
+            project: self,
+        };
+
+        // Install dependencies if specified
+        if !checkout.project().dependencies.is_empty() {
+            tracing::debug!(
+                "Installing {} dependencies for project '{}'...",
+                checkout.project().dependencies.len(),
+                checkout.project().name
+            );
+            let start = std::time::Instant::now();
+            install_dependencies(&checkout)?;
+            tracing::debug!(
+                "Dependency installation completed in {:.2}s",
+                start.elapsed().as_secs_f64()
+            );
+        }
+
+        tracing::debug!("Project setup took: {:.2}s", start.elapsed().as_secs_f64());
+
+        Ok(SetupProject {
+            path: checkout.path,
+            config: checkout.project,
+        })
+    }
+}
+
+struct Checkout<'a> {
+    project: RealWorldProject<'a>,
+    path: PathBuf,
+}
+
+impl<'a> Checkout<'a> {
+    /// Get the virtual environment path
+    fn venv_path(&self) -> PathBuf {
+        self.path.join(".venv")
+    }
+
+    fn project(&self) -> &RealWorldProject<'a> {
+        &self.project
+    }
+}
+
+/// A setup real-world project ready for benchmarking
+pub struct SetupProject<'a> {
```

Comment:
Minor naming suggestion: I used `InstalledProject` for the same concept in a related tool. "setup" has the drawback that the past-tense form is the same, so it's not clear if that project has already been set up, or if it's referring to the actual project setup.